src/execution_tree/primitives/block_operation.cpp

```diff
@@ -20,7 +20,78 @@
 ///////////////////////////////////////////////////////////////////////////////
 namespace phylanx { namespace execution_tree { namespace primitives
 {
+    struct block_operation::step : std::enable_shared_from_this<step>
```

Comment:
This is not how it is supposed to be refactored. In the end the block_operation itself has to be derived from `std::enable_shared_from_this<>`, see your refactoring of the `add_primitive`. The whole point of this refactoring is to avoid having to copy `operands`, `args`, `codename`, and `name` for each of the evaluations on each primitive. If the refactoring is done as proposed here this goal will not be achieved.

Reply:
Oops.. It should be fixed in the new commit.